### ISSUE TITLE
Exposure computation in secret sharer colab

### ISSUE THREAD

**Issue by lxuechen** [tensorflow/privacy#172]
Hi, 

I'm attempting to use the secret sharer utilities that was just released (located [here](https://github.com/tensorflow/privacy/tree/master/tensorflow_privacy/privacy/privacy_tests/secret_sharer)). 

I'm reading the colab [notebook](https://github.com/tensorflow/privacy/blob/master/tensorflow_privacy/privacy/privacy_tests/secret_sharer/secret_sharer_example.ipynb), and I am somewhat confused by the perplexity computation. 

Specifically, I'm referring to this chunk of code in the function `compute_perplexity_for_secret `
```python
  cce = tf.keras.losses.SparseCategoricalCrossentropy(reduction=tf.keras.losses.Reduction.NONE)
  ppls = [{} for _ in range(len(secrets))]
  ppls_reference = [None] * len(secrets)
  for i, secret in enumerate(secrets):
    for r in secret.datasets.keys():
      d = secret.datasets[r]
      loss = np.mean(cce(d['label'], prediction_model.predict(d['data'])),
                     axis=1)
      loss = list(loss)
      if r == 0:
        ppls_reference[i] = loss
      else:
        ppls[i][r] = loss
```

The perplexity variables are directly assigned with the **cross entropy losses**. Clearly, cross entropy and perplexity are entirely different concepts. 

It appears also that this isn't just an issue with the terminology. The **Approximation by distribution model** to exposure in section 4.3 is based on estimating the density of the random variable `Px_theta (s[t])`, where `t` is uniformly drawn from the space `R`. Fitting a skewed normal distribution to the negative logprobs (essentially cross entropy) would result in different distributions compared to fitting one to the actual probabilities values, and this will likely alter the cdf value obtained thereafter. 

Would appreciate some insight on this issue in the example. Thanks in advance.

**Comment by lxuechen** [comment]
Thanks for the quick reply. In the paper, the skewed normal density approximates P[ P_xtheta( s[t] ) = v ], and thus it seems the skewed normal approximation should be fitted using the sampled values of P_xtheta( s[t] ). This quantity is the exponentiated negative cross entropy, and is not the cross entropy per se. The code that I'm confused with is this [line](https://github.com/tensorflow/privacy/blob/caf6f36b80b4f2dd4ff0ea7fcbd6c02054c3d71c/tensorflow_privacy/privacy/privacy_tests/secret_sharer/exposures.py#L75). 

Does that make sense?

**Comment by shs037** [comment]
Sorry I didn't quite get it. I think P_xtheta is the log-perplexity in the paper, and in the code we passed the cross entropy loss to `compute_exposure_extrapolation`. Do you mean we should use perplexity instead of log-perplexity in the function?

**Comment by lxuechen** [comment]
Thanks for the quick response. I think you're right. I misread Pxtheta as the probability. It seems there's also another minor caveat: The logprob is summed across time, as opposed to taking the mean. This would not matter if all canaries are of the same length, which in the present case is indeed true. 

Thanks again.
